libmei/tools/schema.py

```diff
@@ -0,0 +1,235 @@
+            for v in values:
+                if data_type not in self.data_types:
+                    self.data_types[data_type] = []
+                self.data_types[data_type].append(v.get("ident"))
+
+        vallists = self.schema.xpath("//tei:valList[@type='closed' or @type='semi']", namespaces=TEI_RNG_NS)
+        for vl in vallists:
+            element = vl.xpath("./ancestor::tei:classSpec", namespaces=TEI_RNG_NS)
+            if not element:
+                continue
+
+            att_name = vl.xpath("./parent::tei:attDef/@ident", namespaces=TEI_RNG_NS)
+
+            data_list = f"{element[0].get('ident')}@{att_name[0]}"
+            values = vl.xpath(".//tei:valItem", namespaces=TEI_RNG_NS)
+
+            for v in values:
+                if data_list not in self.data_lists:
+                    self.data_lists[data_list] = []
+                self.data_lists[data_list].append(v.get("ident"))
+
+    def invert_attribute_group_structure(self):
+        for module, groups in self.attribute_group_structure.items():
+            for attgroup in groups:
+                self.inverse_attribute_group_structure[attgroup] = module
+
+    def set_active_modules(self):
+        self.active_modules = list(self.element_structure.keys())
+        self.active_modules.sort()
+
+    def __process_att(self, attdef: etree.Element) -> str:
+        """Process attribute definition."""
+        attdef_ident = attdef.get("ident")
+        if "-" in attdef_ident:
+            first, last = attdef_ident.split("-")
+            attdef_ident = f"{first}{last.title()}"
+
+        if attdef.get("ns"):
+            return f"{attdef.get('ns')}|{attdef_ident}"
+        elif ":" in attdef_ident:
+            pfx, att = attdef_ident.split(":")
+            return f"{NAMESPACES[pfx]}|{att}"
+        else:
+            return f"{attdef_ident}"
+
+    def __get_membership(self, member: etree.Element, resarr: list[str]) -> None:
+        """Get attribute groups."""
+        member_attgroup = self.schema.xpath(".//tei:classSpec[@type='atts'][@ident=$nm]", nm=member.get("key"), namespaces=TEI_NS)
+
+        if member_attgroup is None:
+            return None
+
+        member_groupel = member_attgroup[0]
+        if member_groupel.get("ident") == "att.id":
+            return None
+
+        if member_groupel.xpath("./tei:attList/tei:attDef", namespaces=TEI_NS):
+            resarr.append(member_groupel.get("ident"))
+
+        m2s = member_groupel.xpath("./tei:classes/tei:memberOf", namespaces=TEI_NS)
+        for mship in m2s:
+            self.__get_membership(mship, resarr)
+
+    def strpatt(self, string: str) -> str:
+        """Returns a version of the string with any leading att. stripped."""
+        return string.replace("att.", "")
+
+    def strpdot(self, string: str) -> str:
+        """Returns a version of the string without any dots."""
+        return "".join(string.split("."))
+
+    def cc(self, att_name: str) -> str:
```

Comment:
I think "cc_name" would be more speaking

Reply:
Possibly, but I think with the docstring and the pattern of naming other things `_cc` in the cpp.py file, I'll leave it like this for now.